Added multi address slave

Corna opened the pull request:
I've removed the pullups enable in twi.c and I've moved it in Wire.cpp in order to make twi.h/twi.c usable in pure avr-libc projects
I've also added the multi address slave capability; the device can now respond multiple addresses (when configured as slave), and know which one was called thanks to twi_slaveAddress (in twi.c) or TwoWire::slaveAddress (in Wire.cpp). This feature is not supported on uCs without the TWAMR (two wire address mask register), like the ATMega8, and it is disabled with an #ifdef
These boards are:
 - Serial Arduino
 - Arduino USB
 - Arduino Extreme
 - Arduino NG (first version)

Corna commented:
I've committed any requested fix. Can you merge it?

ffissore commented:
Sorry for the late reply @Corna. After 1.6 release, your PR does not merge any more. Can you update it? We'll make the bot build a test IDE then, so to make others test your patch more easily.

Corna commented:
Rebased successfully
I've squashed some commits in order to keep the history clean
The latest Arduino IDE doesn't show the unsupported warning anymore, even if i manually create a 

    #warning Unsupported on this AVR

Is it correct?

matthijskooijman commented:
Warnings not being shown is a known issue, see #1728.

ffissore commented:
@ArduinoBot build this please

Corna commented:
Weird, I've built it successfully on my PC... Do you have the build log?

ffissore commented:
An error in the job configuration. Fixed

ffissore commented:
@ArduinoBot build this please

ffissore commented:
Sorry it looks like I broke our bot

ffissore commented:
@ArduinoBot build this please

ffissore commented:
@ArduinoBot build this please

ffissore commented:
@ArduinoBot build this please

ffissore commented:
@ArduinoBot build this please

Corna commented:
I don't have an Arduino at the moment, but I've previously tested it with an ATmega328P. The ATmega had an address mask of 0x08 (8 addresses) and it was able to manage each one independently.
Can someone test it with an Arduino?
Thanks

ffissore commented:
I've added a new preference "show all compiler warnings" that turns warnings on
Available as nightly builds in the next few hours

raphaelcoeffic commented:
@corna , @ffissore: do you guys still see a possibility to get that merged into mainline? That would be quite useful I think.

ffissore commented:
Sorry you've missed the news, but I don't work for arduino any more. See https://groups.google.com/a/arduino.cc/forum/#!msg/developers/YJLX6AZHem4/kpsLef7XBAAJ

raphaelcoeffic commented:
@ffissore Thx for the reply & good luck where you are now!

raphaelcoeffic commented:
@cmaglie Is there any chance to have this pull request integrated into mainline some time in the future? Thx a lot!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino/Arduino?pullRequest=2403) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino/Arduino?pullRequest=2403) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino/Arduino?pullRequest=2403) it.</sub>
